cpp/src/svm/kernelcache.h

```diff
@@ -0,0 +1,197 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cuda_utils.h>
+#include <linalg/gemm.h>
+#include "cache/cache.h"
+#include "common/cumlHandle.hpp"
+#include "common/host_buffer.hpp"
+#include "gram/grammatrix.h"
+#include "ml_utils.h"
+
+namespace ML {
+namespace SVM {
+
+/**
+ * @brief Collect rows of the training data into contiguous space
+ *
+ * The working set is a subset of all the training examples. Here we collect
+ * all the training vectors that are in the working set.
+ *
+ * @param [in] x training data in column major format, size [n_rows x n_cols]
+ * @param [in] n_rows
+ * @param [in] n_cols
+ * @param [out] x_ws training vectors in the working set in column major format, size [n_ws x n_cols]
+ * @param [in] n_ws the number of elements in the working set
+ * @param [in] ws_idx working set indices (row indices of x), size [n_ws]
+ */
+template <typename math_t>
+__global__ void get_rows(const math_t *x, int n_rows, int n_cols, math_t *x_ws,
+                         int n_ws, const int *ws_idx) {
+  int tid = threadIdx.x + blockIdx.x * blockDim.x;
+  int ws_row = tid % n_ws;  // row idx
+  if (tid < n_ws * n_cols) {
+    int x_row = ws_idx[ws_row];  // row in the original matrix
+    int col = tid / n_ws;        // col idx
+    if (x_row + col * n_rows < n_rows * n_cols) {
+      x_ws[tid] = x[x_row + col * n_rows];
+    }
+  }
+}
+
+/**
+* @brief Buffer to store a kernel tile
+*
+* We calculate the kernel matrix for the vectors in the working set.
+* For every vector x_i in the working set, we always calculate a full row of the
+* kernel matrix K(x_j, x_i), j=1..n_rows.
+*
+* A kernel tile stores all the kernel rows for the working set, i.e. K(x_j, x_i)
+* for all i in the working set, and j in 1..n_rows.
+*
+* The kernel values can be cached to avoid repeated calculation of the kernel
+* function.
+*/
+template <typename math_t>
+class KernelCache {
+ private:
+  const math_t *x;  //!< pointer to the training vectors
+
+  MLCommon::device_buffer<math_t>
+    x_ws;  //!< feature vectors in the current working set
+  MLCommon::device_buffer<int>
+    ws_cache_idx;  //!< cache position of a workspace vectors
+  MLCommon::device_buffer<math_t> tile;  //!< Kernel matrix  tile
+
+  int n_rows;  //!< number of rows in x
+  int n_cols;  //!< number of columns in x
+  int n_ws;    //!< number of elements in the working set
+
+  cublasHandle_t cublas_handle;
+
+  MLCommon::GramMatrix::GramMatrixBase<math_t> *kernel;
```

Comment:
I don't think we need another folder for this but it's up to you. I'd move this whole code to matrix folder in prims.